tensorcircuit/applications/optimization.py

```diff
@@ -0,0 +1,328 @@
+
+        params = opt.update(grads, params)
+        # Update the parameters using the optimizer and gradients.
+
+        if callback is not None:
+            callback(loss, params)
+        # Execute the callback function with the current loss and parameters.
+
+    return params
+    # Return the optimized parameters for the ansatz circuit.
+
+
+def cvar_value(r: list, p: list, percent: float) -> float:
+    """
+    Calculate the Conditional Value at Risk (CVaR) according to the measurement results.
+
+    :param r: The results showing after measurements.
+    :param p: Probabilities corresponding to each result.
+    :param percent: The cut-off percentage of CVaR.
+    :return: The calculated CVaR value.
+    """
+    sorted_indices = np.argsort(r)
+    p = np.array(p)[sorted_indices]
+    r = np.array(r)[sorted_indices]
+
+    sump = 0.0  # The sum of probabilities.
+    count = 0
+    cvar_result = 0.0
+
+    # Iterate over the sorted results and calculate CVaR.
+    while sump < percent:
+        if round(sump + p[count], 6) >= percent:
+            # Add the remaining portion of the last result that exceeds the cut-off percentage.
+            cvar_result += r[count] * (percent - sump)
+            count += 1
+            break
+        else:
+            # Add the entire result to the CVaR calculation.
+            sump += p[count]
+            cvar_result += r[count] * p[count]
+            count += 1
+
+    cvar_result /= percent
+    return cvar_result
+
+def cvar_from_circuit(
+    circuit: Circuit, nsamples: int, Q: Tensor, alpha: float
+) -> float:
+    """
+    Directly calculate the Conditional Value at Risk (CVaR) from a circuit.
+    The CVaR depends on a bunch of measurements.
+
+    :param circuit: The quantum circuit used to prepare the state.
+    :param nsamples: The number of samples to take for measurements.
+    :param Q: The Q-matrix representing the Quadratic Unconstrained Binary Optimization (QUBO) problem.
+    :param alpha: The cut-off percentage for CVaR.
+    :return: The calculated CVaR value.
+    """
+    s = circuit.state()
+    results = tc.quantum.measurement_results(
+        s, counts=nsamples, format="count_dict_bin"
+    )  # Get readouts from the measurements.
+    results = {k: v / nsamples for k, v in results.items()}  # Normalize the results.
+    values = []  # List to store the measurement values.
+    probabilities = []  # List to store the corresponding probabilities.
+
+    # Iterate over the measurement results and calculate the values and probabilities.
+    for k, v in results.items():
+        x = np.array([int(bit) for bit in k])
+        values.append(np.dot(x, np.dot(Q, x)))
+        probabilities.append(v)
+
+    cvar_result = cvar_value(values, probabilities, alpha)
+    # Calculate the CVaR using the cvar_value function.
+
+    return cvar_result
+
+
+def cvar_from_expectation(circuit, Q, alpha: float) -> float:
```

Comment:
This seems like the main difference between this and cvar_from_circuit is that cvar_from_circuit uses empirical measurement results, while cvar_from_expectation classically computes the probabilities of all outcomes. I think we should mention this.